src/controls/Widget.ts

```diff
@@ -107,6 +108,54 @@ export class Widget_markdown implements IRequest<'markdown', Widget_markdown_opt
     get result(): Widget_markdown_output { return this.state }
 }
 
+// 🅿️ custom ==============================================================================
+export type Widget_custom_componentProps_ui = {
+    image: (props: {     
+        size?: string | undefined;
+        img: MediaImageID; 
+    }) => JSX.Element
+}
+export type Widget_custom_componentProps<T = unknown> = {
+    value: undefined | T
+    onChange: (value: undefined | T) => void
+    ui: Widget_custom_componentProps_ui
+}
+
+export type Widget_custom_opts<TViewState>  = ReqInput<{ customComponent: (props: Widget_custom_componentProps<TViewState>) => JSX.Element }>
+export type Widget_custom_serial<TViewState> = StateFields<{ type: 'custom', active: true; componentViewState: undefined | TViewState }>
+export type Widget_custom_state<TViewState>  = StateFields<{ type: 'custom', active: true; componentViewState: undefined | TViewState }>
+export type Widget_custom_output<TViewState> = undefined | TViewState
```

Comment:
I think `undefined` must not be present here, I'd prefer custom widget writers to decide themselves if the state is present or not.

this also means we should probably add an `initialState: () => X` in the input field, so the component is consistently instanciated

Reply:
yeah, this would solve alot of issues

Reply:
I think adding a required default for the initial state would also solve the type inference issues

Reply:
yup. side note: doing it via a lambda function is important to avoid ref sharing or other hard to understand bugs with less experienced devs